### PR TITLE
HackStudio+LibGUI: Improve the git commit experience in HackStudio

### LINKED ISSUE #8979: HackStudio: Application crashes when opening a git diff
When opening the git diff for a file within HackStudio, a crash occurs. This crash has 2 potential causes. (from what I can see so far)

**Cause 1**
This crash only happens if you click the `+` icon beside an unstaged file. The file is staged, but the app crashes.

In ``GitWidget::GitWidget`` line 48, we should be returning if the index is invalid, as when a file is staged, it is removed from ``m_unstaged_files``. 

**Example:**
image

With this fix implemented, it appears that the ``DiffViewer`` has another problem (Cause 2).

**Cause 2**
With the fix for cause 1 implemented, if you try to view the diff of a file, it seems like the ``DiffViewer`` has some out of bounds issue. I haven't been able to get to the bottom of it. I believe it has something to do with ``DiffViewer::paint_event``.
```
HackStudio(32:32): ASSERTION FAILED: i < m_size
../.././AK/Vector.h:150
[#0 HackStudio(32:32)]: Terminating HackStudio(32) due to signal 6
```
(full log [here](https://hst.sh/edufofoqef.php))

### PR DESCRIPTION
**Pull request by cbyrneee**
This pull request improves the git experience in HackStudio by adding a new commit dialog, and fixing a few bugs :^)

**Changes summarized:**
- ``GitRepo`` has been changed to use a singleton (``::the()``) instead of passing around and initializing a bunch of``m_git_repo`` members
- A new commit dialog has been added, which now supports multi-line text input. (preview below) 
- ``on_ctrl_return_pressed`` has been added to ``TextEditor``, in this context it will act as an executor for the Commit button.
- A git commit action has been added to the toolbar. It has a shortcut of ``Ctrl+Alt+P``.
- ``GitWidget``'s existing commit button has been updated to mimic the behavior of the one in the toolbar.
- A crash which would occur when viewing the diff of an unstaged file has been fixed (closes #8979).
- Stage and unstage all buttons have been added to ``GitWidget``.
- A new commit icon has been added (designed by @sin-ack)

**Preview:**
image

### CODE REVIEW

File: Userland/DevTools/HackStudio/Dialogs/Commit/GitCommitDialog.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    m_message_editor->on_change = [this]() {
+        m_commit_button->set_enabled(!m_message_editor->text().is_empty() && on_commit);
+    };
+    m_message_editor->on_ctrl_return_pressed = [this]() {
+        if (m_commit_button->is_enabled())
+            commit();
+    };
+
+    m_message_editor->on_cursor_change = [this]() {
+        m_line_and_column_label->set_text(String::formatted("Line: {}, Col: {}", m_message_editor->cursor().line() + 1, m_message_editor->cursor().column()));
```
Comment (ry-sev) [inline Userland/DevTools/HackStudio/Dialogs/Commit/GitCommitDialog.cpp:38]: I noticed that the line and column label has the format `Line: {}, col: {}`. But the statusbar in the HackStudioWidget has the format `Ln {}, Col: {}`. Obviously, this isn't that important but do you think they should both be the same? Btw I was author of the statusbar so my format may be wrong too.

Comment (cbyrneee) [reply Userland/DevTools/HackStudio/Dialogs/Commit/GitCommitDialog.cpp:38]: I don't personally like the look of ``Ln {}, Col: {}``. I prefer it to really describe what it is, I can change it to be the same if you would like

Comment (ry-sev) [reply Userland/DevTools/HackStudio/Dialogs/Commit/GitCommitDialog.cpp:38]: Its probably fine as is, no need to change it.

**Comment by cbyrneee** [comment]
Everything I touched should use east const now :)

**Comment by cbyrneee** [comment]
Will resolve the merge conflict later

**Comment by awesomekling** [comment]
Looks hella cool. 😎

File: Userland/DevTools/HackStudio/Git/GitRepo.h [hunk 2]
```diff
@@ -1,61 +1,49 @@
...
-
-    static CreateResult try_to_create(const LexicalPath& repository_root);
-    static RefPtr<GitRepo> initialize_repository(const LexicalPath& repository_root);
-
-    Vector<LexicalPath> unstaged_files() const;
-    Vector<LexicalPath> staged_files() const;
+    static GitRepo& the();
+    static void initialize(LexicalPath root);
+
+    bool repo_exists(bool cached);
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/Git/GitRepo.h:19]: To improve readability, let's use an `enum class` here:
```c++
enum class UseCached {
  No,
  Yes
};
bool repo_exists(UseCached);
```

File: Userland/DevTools/HackStudio/Git/GitRepo.cpp [hunk 3]
```diff
@@ -1,131 +1,130 @@
...
 
-    return { CreateResult::Type::Success, adopt_ref(*new GitRepo(repository_root)) };
+GitRepo& GitRepo::the()
+{
+    VERIFY(s_the);
+    return *s_the;
 }
 
-RefPtr<GitRepo> GitRepo::initialize_repository(const LexicalPath& repository_root)
+void GitRepo::initialize(LexicalPath root)
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/Git/GitRepo.cpp:21]: This isn't really related to this PR, but the `GitRepo` class uses `LexicalPath` to pass paths around in many places.
We should refactor the code to use `String`s instead, `LexicalPath` is a 'heavier' object that's meant for path parsing and validation.
But of course this doesn't have to be fixed in this PR.

File: Userland/DevTools/HackStudio/Git/GitRepo.h [hunk 4]
```diff
@@ -1,61 +1,49 @@
...
+    explicit GitRepo(LexicalPath& root)
+        : m_root(root)
     {
     }
 
-    Vector<LexicalPath> modified_files() const;
-    Vector<LexicalPath> untracked_files() const;
-
-    String command(const Vector<String>& command_parts) const;
+    Vector<LexicalPath> parse_files_list(const String& raw_result);
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/Git/GitRepo.h:45]: Any reason this was changed to not be `static` anymore?

File: Userland/DevTools/HackStudio/Git/GitWidget.cpp [hunk 5]
```diff
@@ -133,13 +134,18 @@ void GitWidget::unstage_file(const LexicalPath& file)
 
 void GitWidget::commit()
```
Comment (itamar8910) [inline Userland/DevTools/HackStudio/Git/GitWidget.cpp:171]: A bit of a nitpick but the logic here could be nicer if we flipped the if/else logic and did early return.

**Review by itamar8910 — COMMENTED** [review]
Very nice improvements :+1: 
Please see my comments.

**Comment by itamar8910** [comment]
Also note that there are some minor conflicts with the main branch.

**Comment by cbyrneee** [comment]
Thank you, yes I am aware of those and will fix it soon. Been busy the past few days :p

**Pull request closed without merging**